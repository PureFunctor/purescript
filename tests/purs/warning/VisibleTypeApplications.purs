-- @shouldWarnWith ShadowedTypeVar
-- @shouldWarnWith MissingTypeDeclaration
-- @shouldWarnWith MissingTypeDeclaration
-- @shouldWarnWith MissingTypeDeclaration
-- @shouldWarnWith MissingTypeDeclaration
module Main where

-- Application on a wildcard should remove an abstraction
const :: forall @a @b. a -> b -> a
const a _ = a

-- const' :: forall a @b. a -> b -> a
const' = const @_


-- Application on a wildcard should generalize with a constraint
class MultiKind :: (Type -> Type -> Type) -> Constraint
class MultiKind @f where
  foo :: forall @a @b. f a b -> String

data Tuplet1 a b c = Tuplet1 a b c
data Tuplet2 a b c = Tuplet2 a b c

instance MultiKind (Tuplet1 a) where
  foo _ = "Tuplet1"

instance MultiKind (Tuplet2 a) where
  foo _ = "Tuplet2"

testFooGeneralize = foo @_ @Int @Int


-- Only top-level foralls should be vta-ified
data A :: Type -> Type
data A a = A (forall a. a -> a)

_A = A

class F :: Type -> Constraint
class F @a where
  f :: a -> (forall a. a -> a)

_f = f
