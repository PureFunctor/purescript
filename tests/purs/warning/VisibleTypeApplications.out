Warning 1 of 5:

  in module [33mMain[0m
  at tests/purs/warning/VisibleTypeApplications.purs:41:3 - 41:31 (line 41, column 3 - line 41, column 31)

    Type variable [33ma[0m was shadowed.

  in type declaration for [33mf[0m

  See https://github.com/purescript/documentation/blob/master/errors/ShadowedTypeVar.md for more information,
  or to contribute content related to this warning.

Warning 2 of 5:

  in module [33mMain[0m
  at tests/purs/warning/VisibleTypeApplications.purs:30:1 - 30:37 (line 30, column 1 - line 30, column 37)

    No type declaration was provided for the top-level declaration of [33mtestFooGeneralize[0m.
    It is good practice to provide type declarations as a form of documentation.
    The inferred type of [33mtestFooGeneralize[0m was:
    [33m                                                    [0m
    [33m  forall f76. MultiKind f76 => f76 Int Int -> String[0m
    [33m                                                    [0m

  in value declaration [33mtestFooGeneralize[0m

  See https://github.com/purescript/documentation/blob/master/errors/MissingTypeDeclaration.md for more information,
  or to contribute content related to this warning.

Warning 3 of 5:

  in module [33mMain[0m
  at tests/purs/warning/VisibleTypeApplications.purs:13:1 - 13:18 (line 13, column 1 - line 13, column 18)

    No type declaration was provided for the top-level declaration of [33mconst'[0m.
    It is good practice to provide type declarations as a form of documentation.
    The inferred type of [33mconst'[0m was:
    [33m                                   [0m
    [33m  forall a84 b85. a84 -> b85 -> a84[0m
    [33m                                   [0m

  in value declaration [33mconst'[0m

  See https://github.com/purescript/documentation/blob/master/errors/MissingTypeDeclaration.md for more information,
  or to contribute content related to this warning.

Warning 4 of 5:

  in module [33mMain[0m
  at tests/purs/warning/VisibleTypeApplications.purs:43:1 - 43:7 (line 43, column 1 - line 43, column 7)

    No type declaration was provided for the top-level declaration of [33m_f[0m.
    It is good practice to provide type declarations as a form of documentation.
    The inferred type of [33m_f[0m was:
    [33m                                           [0m
    [33m  forall @a. F a => a -> (forall a. a -> a)[0m
    [33m                                           [0m

  in value declaration [33m_f[0m

  See https://github.com/purescript/documentation/blob/master/errors/MissingTypeDeclaration.md for more information,
  or to contribute content related to this warning.

Warning 5 of 5:

  in module [33mMain[0m
  at tests/purs/warning/VisibleTypeApplications.purs:37:1 - 37:7 (line 37, column 1 - line 37, column 7)

    No type declaration was provided for the top-level declaration of [33m_A[0m.
    It is good practice to provide type declarations as a form of documentation.
    The inferred type of [33m_A[0m was:
    [33m                                         [0m
    [33m  forall a88. (forall a. a -> a) -> A a88[0m
    [33m                                         [0m

  in value declaration [33m_A[0m

  See https://github.com/purescript/documentation/blob/master/errors/MissingTypeDeclaration.md for more information,
  or to contribute content related to this warning.

